pkg/kubelet/api/testing/fake_runtime_service.go

```diff
@@ -19,66 +19,356 @@ package testing
+				continue
+			}
+		}
+
+		result = append(result, &runtimeApi.PodSandbox{
+			Id:        s.Id,
+			Name:      s.Name,
+			State:     s.State,
+			CreatedAt: s.CreatedAt,
+			Labels:    s.Labels,
+		})
+	}
+
+	return result, nil
 }
 
-func (r *fakeRuntimeService) CreateContainer(podSandboxID string, config *runtimeApi.ContainerConfig, sandboxConfig *runtimeApi.PodSandboxConfig) (string, error) {
-	return "", fmt.Errorf("not implemented")
+func (r *FakeRuntimeService) CreateContainer(podSandboxID string, config *runtimeApi.ContainerConfig, sandboxConfig *runtimeApi.PodSandboxConfig) (string, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "CreateContainer")
+
+	// ContainerID should be randomized for real container runtime, but here just use
+	// container's name for easily making fake containers.
+	containerID := config.GetName()
+	createdAt := time.Now().Unix()
+	createdState := runtimeApi.ContainerState_CREATED
+	imageRef := config.Image.GetImage()
+	r.Containers[containerID] = &FakeContainer{
+		ContainerStatus: runtimeApi.ContainerStatus{
+			Id:          &containerID,
+			Name:        config.Name,
+			Image:       config.Image,
+			ImageRef:    &imageRef,
+			CreatedAt:   &createdAt,
+			State:       &createdState,
+			Labels:      config.Labels,
+			Annotations: config.Annotations,
+		},
+		SandboxID: podSandboxID,
+	}
+
+	return containerID, nil
 }
 
-func (r *fakeRuntimeService) StartContainer(rawContainerID string) error {
-	return fmt.Errorf("not implemented")
+func (r *FakeRuntimeService) StartContainer(rawContainerID string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "StartContainer")
+
+	c, ok := r.Containers[rawContainerID]
+	if !ok {
+		return fmt.Errorf("container %s not found", rawContainerID)
+	}
+
+	// Set container to running.
+	startedAt := time.Now().Unix()
+	runningState := runtimeApi.ContainerState_RUNNING
+	c.State = &runningState
+	c.StartedAt = &startedAt
+
+	return nil
 }
 
-func (r *fakeRuntimeService) StopContainer(rawContainerID string, timeout int64) error {
-	return fmt.Errorf("not implemented")
+func (r *FakeRuntimeService) StopContainer(rawContainerID string, timeout int64) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "StopContainer")
+
+	c, ok := r.Containers[rawContainerID]
+	if !ok {
+		return fmt.Errorf("container %s not found", rawContainerID)
+	}
+
+	// Set container to exited state.
+	finishedAt := time.Now().Unix()
+	exitedState := runtimeApi.ContainerState_EXITED
+	c.State = &exitedState
+	c.FinishedAt = &finishedAt
+
+	return nil
 }
 
-func (r *fakeRuntimeService) RemoveContainer(rawContainerID string) error {
-	return fmt.Errorf("not implemented")
+func (r *FakeRuntimeService) RemoveContainer(rawContainerID string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "RemoveContainer")
+
+	if _, ok := r.Containers[rawContainerID]; ok {
+		delete(r.Containers, rawContainerID)
```

Comment:
nit: I don't think you need to check whether the key exists in the map or not. Just delete it.